File: problem_050.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "swiftc problem_050.swift -o problem_050" -*-
// Copyright (c) 2016 Michael Caldwell

print("Project Euler - Problem 50:")
print("Which prime, below one-million, can be written as the sum of the most consecutive primes?\n")

let limit = 1000000
var primeSieve = [Int](count: limit, repeatedValue: 1)
primeSieve[0] = 0
primeSieve[1] = 0
var primes:[Int] = []

// Populate primes list
for n in 2...limit-1 {
    if primeSieve[n] == 1 {
        primes.append(n)
        for var i=2; i*n < primeSieve.count; i++ {
            primeSieve[n*i] = 0
        }
    }
}

struct ConsecutivePrimes {
    var count = 0
    var value = 0
}

var max = ConsecutivePrimes()
for i in 0...100 {
    var current = ConsecutivePrimes()
    for p in i...primes.count-1 {
        current.value += primes[p]
        current.count += 1

        if current.value >= limit {
            break
        }
        if primeSieve[current.value] == 1 {
            if current.count > max.count {
                max.count = current.count
                max.value = current.value
            }
        }
    }
}

print(max.value)

File: problem_050.rs
// -*- compile-command: "rustc -o problem_050_rs problem_050.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn is_prime(n: u64) -> bool {
    if n < 2 { return false; }
    if n == 2 { return true; }
    if n % 2 == 0 { return false; }

    let limit = (n as f64).sqrt() as u64;
    for i in 3..(limit+1) {
        if n % i == 0 { return false; }
    }
    return true;
}

fn main() {
    println!("Project Euler - Problem 50");
    println!("Which prime, below one-million, can be written as the sum of the most consecutive primes?\n");

    let mut max_sum = 0;
    let mut max_count = 0;

    for start in 1..27 {
        let mut sum = 0;
        let mut count = 0;
        for n in start..4000 {
            if is_prime(n) {
                sum += n;
                count += 1;
            }

            if sum > 1000000 { break; }

            if is_prime(sum) {
                if count > max_count {
                    //println!("Candidate: {} {} {} {}", sum, count, start, n);
                    max_count = count;
                    max_sum = sum;
                }
            }
        }
    }

    println!("{}", max_sum);
}
